components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java

```diff
@@ -0,0 +1,1486 @@
+
+        if (dataSourceMap.get(database) != null) {
+            return dataSourceMap.get(database).getConnection();
+        }
+        throw new RuntimeException("No datasource initiated for database: " + database);
+    }
+
+    @BeforeMethod
+    public void setup() throws Exception {
+
+        idPManagementDAO = new IdPManagementDAO();
+        cacheBackedIdPMgtDAO = new CacheBackedIdPMgtDAO(idPManagementDAO);
+        initiateH2Database(DB_NAME, getFilePath("h2.sql"));
+        mockStatic(IdentityTenantUtil.class);
+        when(IdentityTenantUtil.getTenantDomain(anyInt())).thenReturn(TENANT_DOMAIN);
+    }
+
+    @AfterMethod
+    public void tearDown() throws Exception {
+
+        closeH2Database();
+    }
+
+    @DataProvider
+    public Object[][] getIdPsData() {
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, 2},
+                {SAMPLE_TENANT_ID2, TENANT_DOMAIN, 1},
+                {NOT_EXISTING_TENANT_ID, TENANT_DOMAIN, 0}
+        };
+    }
+
+    @Test(dataProvider = "getIdPsData")
+    public void testGetIdPs(int tenantId, String tenantDomain, int idpCount) throws Exception {
+
+        mockStatic(IdentityDatabaseUtil.class);
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+
+            // Directly providing initiated db connection as connection parameter
+            List<IdentityProvider> idpList1 = cacheBackedIdPMgtDAO.getIdPs(connection, tenantId, tenantDomain);
+            assertEquals(idpList1.size(), idpCount);
+            // Providing null as connection parameter so value of connection parameter will be retrieved
+            // from the mock above
+            List<IdentityProvider> idpList2 = cacheBackedIdPMgtDAO.getIdPs(null, tenantId, tenantDomain);
+            assertEquals(idpList2.size(), idpCount);
+        }
+    }
+
+    @Test(dataProvider = "getIdPsData")
+    public void testGetIdPsException(int tenantId, String tenantDomain, int idpCount) throws Exception {
+
+        mockStatic(IdentityDatabaseUtil.class);
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+        }
+        assertThrows(IdentityProviderManagementException.class, () ->
+                cacheBackedIdPMgtDAO.getIdPs(null, tenantId, tenantDomain));
+    }
+
+    @DataProvider
+    public Object[][] getIdPsSearchData() {
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, "testIdP1", 1},
+                {SAMPLE_TENANT_ID2, TENANT_DOMAIN, "testIdP3", 1},
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, "", 2},
```

Comment:
All scenarios getting same tenant domain? Shall we use different tenant domain combination as well?

Reply:
Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa